### PR TITLE
support-ranges

### PR DESCRIPTION
**Pull request by AlexCue987**
* partially resolve the following issue: https://github.com/kotest/kotest/issues/3726
* only ranges in this PR, Iterables will be dealt with in the next one
* did not update versioned_docs, as this is a new feature not in those versions.

### CODE REVIEW

File: documentation/docs/assertions/ranges.md [hunk 1]
```diff
@@ -0,0 +1,14 @@
+---
+id: ranges
+title: Range Matchers
+slug: range-matchers.html
+sidebar_label: Ranges
+---
+
+This page describes the rich assertions (matchers) that are available for ClosedRange and OpenEndRange types.
```
Comment (LeoColman) [inline documentation/docs/assertions/ranges.md]:
```suggestion
This page describes the rich assertions (matchers) that are available for [ClosedRange](https://kotlinlang.org/docs/ranges.html) and [OpenEndRange](https://kotlinlang.org/docs/ranges.html) types.
```

File: gradle.properties [hunk 2]
```diff
@@ -36,3 +36,5 @@ kotlin.incremental=true
 kotlin.incremental.js=true
 
 org.gradle.caching=true
+
+jvmOnly=true
```
Comment (LeoColman) [inline gradle.properties]: I don't think we need this change commited. Could you remove it?

File: kotest-assertions/kotest-assertions-core/src/jvmMain/kotlin/io/kotest/matchers/ranges/bein.kt [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+ * Assertion to check that this element is in [range]. This assertion checks by value, and not by reference,
+ * therefore even if the exact instance is not in [range] but another instance with same value is present, the
+ * test will pass.
+ *
+ * An empty range will always fail. If you need to check for empty range, use [Iterable.shouldBeEmpty]
+ *
+ */
+fun <T: Comparable<T>> beIn(range: ClosedRange<T>) = object : Matcher<T> {
+   override fun test(value: T): MatcherResult {
+      if (range.isEmpty())    throw AssertionError("Asserting content on empty range. Use Iterable.shouldBeEmpty() instead.")
```
Comment (LeoColman) [inline kotest-assertions/kotest-assertions-core/src/jvmMain/kotlin/io/kotest/matchers/ranges/bein.kt]:
```suggestion
      if (range.isEmpty()) throw AssertionError("Asserting content on empty range. Use Iterable.shouldBeEmpty() instead.")
```

File: kotest-assertions/kotest-assertions-core/src/jvmMain/kotlin/io/kotest/matchers/ranges/bein.kt [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+
+      val match = value in range
+
+      return MatcherResult(
+         match,
+         { "Range should contain ${value.print().value}, but doesn't. Possible values: ${range.print().value}" },
+         {
+            "Range should not contain ${value.print().value}, but does. Forbidden values: ${range.print().value}"
+         })
```
Comment (LeoColman) [inline kotest-assertions/kotest-assertions-core/src/jvmMain/kotlin/io/kotest/matchers/ranges/bein.kt]:
```suggestion
         { "Range should not contain ${value.print().value}, but does. Forbidden values: ${range.print().value}" }
     )
```

**Review by LeoColman — CHANGES_REQUESTED** [review]
Amazing PR!

Suggested a few changes, being **remove gradle.properties from commit the single blocker**

**Comment by LeoColman** [comment]
Also, I think ranges are supported in Kotlin Common. Can we move this assertion to that package?

**Comment by AlexCue987** [comment]
> Also, I think ranges are supported in Kotlin Common. Can we move this assertion to that package?

Can you please elaborate? Ranges themselves are in `package kotlin.ranges`. I've searched on `ClosedRange` in this repo and all I found was in `package io.kotest.matchers.bigdecimal`  - `fun beInClosedRange(range: ClosedRange<BigDecimal>)`. What am I doing wrong?

File: gradle.properties (hunk 2 above)
Comment (AlexCue987) [reply gradle.properties]: fixed. Sorry about that.

**Comment by LeoColman** [comment]
BigDecimals are only available in Java, so it makes sense to put it only in JVM

Ranges, however, are part of the common library: https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.ranges/-closed-range/ 


For this reason, we don't need to limit to the 'jvm' module, we can move it up to 'commonMain', and users will be able to use these functions in `kotlin-js`, `kotlin-native`, etc

**Comment by AlexCue987** [comment]
> BigDecimals are only available in Java, so it makes sense to put it only in JVM
> 
> Ranges, however, are part of the common library: https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.ranges/-closed-range/
> 
> For this reason, we don't need to limit to the 'jvm' module, we can move it up to 'commonMain', and users will be able to use these functions in `kotlin-js`, `kotlin-native`, etc

Got it, TY, will do!

**Comment by AlexCue987** [comment]
> BigDecimals are only available in Java, so it makes sense to put it only in JVM
> 
> Ranges, however, are part of the common library: https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.ranges/-closed-range/
> 
> For this reason, we don't need to limit to the 'jvm' module, we can move it up to 'commonMain', and users will be able to use these functions in `kotlin-js`, `kotlin-native`, etc

Done

**Review by LeoColman — APPROVED** [review]
Very nice PR!

**Comment by AlexCue987** [comment]
> Very nice PR!

Thanks a lot!

**Comment by AlexCue987** [comment]
@LeoColman can you help me understand failing api_check step? TIA!

**Comment by kshired** [comment]
@AlexCue987 

You have to dump api like this `./gradlew apiDump`

**Comment by AlexCue987** [comment]
> @AlexCue987
> 
> You have to dump api like this `./gradlew apiDump`

Done - TY!

**Pull request merged**